File: brain-move/games/passi_ricorda.py
import random
import time
from games.base_game import BaseGame


class PassiRicorda(BaseGame):
    """Gioco 2: Memoria + movimento - camminata sul posto e richiamo sequenza."""

    ITEMS = ["mela", "chiave", "fiore", "casa", "sole", "tazza", "libro", "stella"]
    ITEM_ICONS = {
        "mela": "\U0001F34E", "chiave": "\U0001F511", "fiore": "\U0001F33C",
        "casa": "\U0001F3E0", "sole": "\U0001F31E", "tazza": "\u2615",
        "libro": "\U0001F4D6", "stella": "\u2B50",
    }

    def __init__(self, difficulty_engine, scoring_engine):
        super().__init__(difficulty_engine, scoring_engine)
        self.name = "passi_ricorda"
        self.display_name = "Passi e Ricorda"
        self.sequence = []
        self.user_answer = []
        self.phase = "showing"
        self.show_index = 0
        self.show_timer = 0
        self.show_interval = 2.0
        self.recall_options = []
        self.step_count = 0
        self.movement_valid = False
        self.correct_count = 0
        self.total_attempts = 0

    def start(self):
        super().start()
        level = self.difficulty.get_level()
        seq_len = min(2 + level, len(self.ITEMS))
        self.sequence = random.sample(self.ITEMS, seq_len)
        self.user_answer = []
        self.recall_options = list(self.ITEMS)
        random.shuffle(self.recall_options)
        self.phase = "showing"
        self.show_index = 0
        self.show_timer = time.time()
        self.step_count = 0
        self.correct_count = 0
        self.total_attempts = 0
        self.feedback_message = "Guarda la sequenza e cammina sul posto"

    def update(self, pose_data=None, dt=0):
        if not self.running:
            return

        if self.phase == "showing":
            elapsed = time.time() - self.show_timer
            if elapsed >= self.show_interval:
                self.show_index += 1
                self.show_timer = time.time()
                if self.show_index >= len(self.sequence):
                    self.phase = "moving"
                    self.feedback_message = "Ora cammina piano sul posto!"
                    self.movement_timer = time.time()
        elif self.phase == "moving":
            move_elapsed = time.time() - self.movement_timer
            if move_elapsed > 8:
                self.phase = "recall"
                self.feedback_message = "Scegli gli elementi nell'ordine giusto"
        elif self.phase == "recall":
            pass

    def select_item(self, item):
        if self.phase != "recall":
            return
        self.user_answer.append(item)
        if len(self.user_answer) == len(self.sequence):
            self._check_answer()

    def _check_answer(self):
        self.total_attempts += 1
        correct = self.user_answer == self.sequence
        if correct:
            self.correct_count += 1
            self.scoring.add_score(self.name, 20)
            self.feedback_message = "Perfetto, sequenza corretta!"
        else:
            self.feedback_message = "Non preoccuparti, riproviamo"

        if self.total_attempts >= 3:
            self.stop()
            accuracy = self.correct_count / self.total_attempts
            self.result = {
                "completed": True,
                "accuracy": accuracy,
                "correct": self.correct_count,
                "total": self.total_attempts,
                "sequence": self.sequence,
            }
            self.difficulty.adapt(accuracy)

    def get_current_item(self):
        if self.phase == "showing" and self.show_index < len(self.sequence):
            return self.sequence[self.show_index]
        return None

    def get_options(self):
        return self.recall_options

    def get_progress(self):
        if self.phase == "showing":
            return self.show_index / max(len(self.sequence), 1) * 0.3
        elif self.phase == "moving":
            return 0.3 + 0.3 * min((time.time() - self.movement_timer) / 8, 1)
        elif self.phase == "recall":
            return 0.6 + 0.4 * (len(self.user_answer) / max(len(self.sequence), 1))
        return 0

File: brain-move-web/src/games/passi-ricorda.ts
import { BaseGame } from './base-game'
import { Difficulty } from '../engine/difficulty'
import { Scoring } from '../engine/scoring'

const ITEMS = ["mela", "chiave", "fiore", "casa", "sole", "tazza", "libro", "stella"]
const ITEM_ICONS: Record<string, string> = {
  mela: "\u{1F34E}", chiave: "\u{1F511}", fiore: "\u{1F33C}", casa: "\u{1F3E0}",
  sole: "\u{1F31E}", tazza: "\u2615", libro: "\u{1F4D6}", stella: "\u2B50",
}

export class PassiRicorda extends BaseGame {
  sequence: string[] = []
  userAnswer: string[] = []
  phase: "showing" | "moving" | "recall" | "feedback" = "showing"
  showIndex = 0
  showTimer = 0
  showInterval = 2.0
  recallOptions: string[] = []
  movementTimer = 0
  correctCount = 0
  totalAttempts = 0
  itemResults: Array<"correct" | "wrong" | null> = []
  feedbackTimer = 0
  lastAttemptCorrect = false
  lastItemResult: "correct" | "wrong" | null = null
  lastItemValue: string | null = null

  constructor(difficulty: Difficulty, scoring: Scoring) {
    super(difficulty, scoring)
    this.name = "passi_ricorda"
    this.displayName = "Passi e Ricorda"
  }

  start(): void {
    super.start()
    const level = this.difficulty.getLevel()
    const seqLen = Math.min(2 + level, ITEMS.length)
    this.sequence = [...ITEMS].sort(() => Math.random() - 0.5).slice(0, seqLen)
    this.userAnswer = []
    this.recallOptions = [...ITEMS].sort(() => Math.random() - 0.5)
    this.phase = "showing"
    this.showIndex = 0
    this.showTimer = Date.now()
    this.correctCount = 0
    this.totalAttempts = 0
    this.itemResults = []
    this.lastAttemptCorrect = false
    this.lastItemResult = null
    this.lastItemValue = null
    this.feedbackMessage = "Guarda la sequenza e cammina sul posto"
  }

  update(_poseData?: any, dt = 0): void {
    if (!this.running) return
    if (this.phase === "showing") {
      const elapsed = (Date.now() - this.showTimer) / 1000
      if (elapsed >= this.showInterval) {
        this.showIndex++
        this.showTimer = Date.now()
        if (this.showIndex >= this.sequence.length) {
          this.phase = "moving"
          this.movementTimer = Date.now()
          this.feedbackMessage = "Ora cammina piano sul posto!"
        }
      }
    } else if (this.phase === "moving") {
      if ((Date.now() - this.movementTimer) / 1000 > 8) {
        this.phase = "recall"
        this.feedbackMessage = "Scegli gli elementi nell'ordine giusto"
      }
    } else if (this.phase === "feedback") {
      this.feedbackTimer -= dt
      if (this.feedbackTimer <= 0) {
        if (this.lastAttemptCorrect || this.totalAttempts >= 3) {
          this.stop()
          const accuracy = this.correctCount / this.totalAttempts
          this.result = { completed: true, accuracy, correct: this.correctCount, total: this.totalAttempts, sequence: this.sequence }
          this.difficulty.adapt(accuracy)
        } else {
          this.userAnswer = []
          this.itemResults = []
          this.phase = "recall"
          this.lastItemResult = null
          this.lastItemValue = null
          this.feedbackMessage = "Riprova! Scegli gli elementi nell'ordine giusto"
        }
      }
    }
  }

  selectItem(item: string): "correct" | "wrong" | null {
    if (this.phase !== "recall") return null
    const idx = this.userAnswer.length
    const result: "correct" | "wrong" = item === this.sequence[idx] ? "correct" : "wrong"
    this.lastItemResult = result
    this.lastItemValue = item
    this.userAnswer.push(item)
    this.itemResults.push(result)
    if (this.userAnswer.length === this.sequence.length) this._checkAnswer()
    return result
  }

  private _checkAnswer(): void {
    this.totalAttempts++
    this.lastAttemptCorrect = this.itemResults.every(r => r === "correct")
    if (this.lastAttemptCorrect) {
      this.correctCount++
      this.scoring.addScore(this.name, 20)
      this.feedbackMessage = "Perfetto, sequenza corretta!"
    } else {
      this.feedbackMessage = "Non preoccuparti, riproviamo"
    }
    this.phase = "feedback"
    this.feedbackTimer = 2.0
    this.lastItemResult = null
    this.lastItemValue = null
  }

  getCurrentItem(): string | null {
    if (this.phase === "showing" && this.showIndex < this.sequence.length)
      return this.sequence[this.showIndex]
    return null
  }

  getOptions(): string[] { return this.recallOptions }
  getItemIcon(item: string): string { return ITEM_ICONS[item] ?? "?" }
  getItemResult(idx: number): "correct" | "wrong" | null { return this.itemResults[idx] ?? null }

  getProgress(): number {
    if (this.phase === "showing") return (this.showIndex / Math.max(this.sequence.length, 1)) * 0.3
    if (this.phase === "moving") return 0.3 + 0.3 * Math.min((Date.now() - this.movementTimer) / 8000, 1)
    if (this.phase === "recall" || this.phase === "feedback") return 0.6 + 0.4 * (this.userAnswer.length / Math.max(this.sequence.length, 1))
    return 0
  }
}
